lib/elixir/lib/gen_server.ex

```diff
@@ -748,7 +748,16 @@ defmodule GenServer do
   """
   @spec stop(server, reason :: term, timeout) :: :ok
   def stop(server, reason \\ :normal, timeout \\ :infinity) do
-    :gen.stop(server, reason, timeout)
+    case whereis(server) do
+      nil ->
+        exit(:noproc)
+
+      pid when pid == self() ->
+        exit({:calling_self, {__MODULE__, :stop, [server, reason, timeout]}})
+
+      pid ->
+        :proc_lib.stop(pid, reason, timeout)
```

Comment:
You will also have to wrap this in a try/catch and reraise any exit with the module+function+args. See how `call/3` in this same module does it.